new installation mode: isolated-mode (#436)
npm/rfcs: accepted/0042-isolated-mode.md

# Isolated mode

## Summary

This RFC is a proposal to add a new opt-in installation mode.

This mode aims at preventing workspaces from over-sharing their dependencies. With this mode, two workspaces will share a dependency only if both have declared it in their package.json.

> This RFC aims at improving projects with workspaces, but the suggested implementation will also happen to bring value to projects not using workspaces. Because `isolated-mode` is valuable for any project, it will also be available for projects not using workspaces.

Reviewer: This is bit repetitive.

While this RFC aims at improving projects with workspaces, `isolated-mode` could be valuable for any project, whether or not it uses workspaces.

## Motivation

The introduction of workspaces in npm v7 brought the ability to split large code-bases into smaller pieces of code with declared dependencies between them.

Several build tools took advantage of these declared dependencies to split the builds into multiple steps (one per workspace). These build tools take advantage of the declared dependencies between workspaces to only rebuild the workspaces that have not changed since the last build. With the current way npm works, any dependency change to any workspace may affect any other workspace. This means that the build systems need to rebuild the entire project every time a single workspace modifies a dependency. This is inefficient and does not scale.

> 

Reviewer: We should mention that workspaces in npm v7 facilitate the "monorepo" pattern (since you mention it later).

The motivations are:
* monorepo performance
* monorepo dependency isolation (prevent side effects when changing dependencies of one workspace to another)

I think this could be a bit more direct and clearer in this section specifically.

## Naming

This mode is called `isolated-mode` because it makes the boundaries between packages and workspaces stronger, making them more isolated. 

To make the discussions more efficient, we could also name the current default installation mode of npm. We will call this mode `hoisted-mode` as packages are shared across workspaces by hoisting them to the root of the project. This term is already known by developers familiar with yarn or pnpm.

## More on the problem

Packages and workspaces declare their dependencies to npm by using special fields in their `package.json`. The Node.js runtime is unaware of these special fields, instead it uses a [module resolution algorithm](https://nodejs.org/api/modules.html) to determine the meaning of a module importing another module. npm bridges this gap by converting the declared dependency graph into a folder structure which makes sense to Node.js. This operation is call `reification`.

> When converting the dependency graph to a folder structure, `hoisted-mode` loses information. Multiple different dependency graphs can be converted to an identical folder structure. Since the folder structure is all that Node.js uses to resolve modules, Node.js misses some information about dependencies. This loss of information is most frequently visible through the fact that workspaces are able to successfully import the dependencies declared by other workspaces.

Reviewer 2: ```suggestion
When converting the dependency graph to a folder structure, `hoisted-mode` loses information. Multiple different dependency graphs can be converted to an identical folder structure. Since the folder structure is all that Node.js uses to resolve modules, Node.js misses some information about dependencies. This loss of information is most frequently visible through the fact that packages in the tree are able to successfully import packages that they have not explicitly declared as dependencies.
```

### Forgetting to declare dependencies

When workspaces can successfully use code of a package without having a dependency on it, people forget to declare their dependencies. This leads to situations where updating the dependencies of one workspace breaks a seemingly unrelated workspace.

> It is worth noting that static code analysis tools can help significantly reduce the frequency of these mistakes. While static code analysis tools are probably sufficient for small projects, they are not for large projects. "isolated-mode" not only helps get the dependencies right but also _proves_ that the dependencies are correct. This proof is essential to build trustable build systems which use caching, scoped installations.

Reviewer 3: Why are they not for large projects? I’ve used many extremely large projects with eslint-plugin-import and it works just fine.

Author: Static analysis tools are not enough guarantee to be able to have a build system reliably do caching  (not re-building everything when only a small part of the repo changes). This is because a dependency of one workspace can break another unrelated workspace. Static code analysis does not remove this risk.

Reviewer 3: In my experience, static code analysis tools are sufficient to force all dependencies to be explicitly declared, in large as well as small projects, which seems like it obviates the problem you're describing.

Author: one example that we hit is a workspace adding a dependency to `@types/jasmine` and that breaks all the other workspaces which use `@types/jest`. Because `@types/jasmine` will be hoisted and included in the compilation of all the workspaces, making them fail. This cannot be caught by code analysis tools because the dependencies are actually correctly declared.

Reviewer 3: ok, that makes sense - maybe instead of "they are not", a more nuanced sentence could be used, since it's really not the size of the project that causes the issue, it's "there's more than one child workspace".

It also might be unique to packages with TS types, or to these specific packages which have colliding type definitions - do you have any examples of this happening with JS?

Author: another example:
- `workspace 1` depends on `foo`
- `foo` requires `yargs` but does not declare it as a dependency
- `workspace 2` depends on `yargs`
I all works fine, foo can require `yargs` because it is hoisted.

- now `workspace 2` get rid of `yargs`
- `workspace 1` is broken because `foo` cannot import `yargs` anymore

In this situation all the dependencies are correctly declared in my repo, so it would not be caught by code analysis.

Author: > maybe instead of "they are not", a more nuanced sentence could be used, since it's really not the size of the project that causes the issue, it's "there's more than one child workspace".

It is not limited to repositories using workspaces, in the example above, if you replace `workspace 2` by an external package called `bar` and you make `workspace 1` take a dependency on `bar` you get the same issue.

It is actually not limited to the size of the repo, but the likelihood of this being an issue increases with the complexity of the dependency graph, to a point a which it becomes unmanageable passed a certain scale (which we have passed in internal repos in Microsoft).

Reviewer 3: Sure - but the likelihood isn’t zero in a small repo either, so the size of the repo doesn’t flip a binary, it instead increases a probability - and it’d be great for the prose here to reflect that.

> ### Duplication

Reviewer 3: Is this something orthogonal? iow, this seems like something worth solving for both modes, not just for this one.

Author: @ljharb can you clarify what would be worth solving for both modes?

Reviewer 3: In the default mode, could you not still use symlinking to share package versions across projects and save disk space?

Author: It is difficult to use symlinks without fully implementing a store like proposed in this RFC. The reason is that when you require() or require.resolve(), you always get back a real path, so you look the context of which you come from (unless you have that context encoded in the store like this RFC suggests).

Another option for using symlinks is to use the preserve symlink node option, but it comes with bad implications like bad perf because a module can be instanciated many times.
 https://nodejs.org/api/cli.html#cli_preserve_symlinks

Reviewer 2: We do use symlinks in some cases to resolve otherwise impossible scenarios.  For example, in the "pathological nesting loop" scenario, we use symlinks to avoid an infinite regress.  The simplest formulation of that scenario is:

```
x@1 -> (y@1)
x@2 -> (y@2)
y@1 -> (x@2)
y@2 -> (x@1)
```

When reifying, the first time we notice that a package is descended from a package of the same name that does not match the dep being installed, which is in turn descended from a package of the same name which _does_ match the dep being installed, we place a symbolic link instead of unpacking it a second time.

So instead of this:

```
root
+-- x@1
+-- y@1
    +-- x@2
    +-- y@2
        +-- x@1
        +-- y@1
            +-- x@2 ...
```

we have this instead:

```
root
+-- x@1
+-- y@1
    +-- x@2
    +-- y@2
        +-- x@1
        +-- y -> root/node_modules/y
```

I have explored a "dedupe harder" mode.  In this mode, we walk the nodes in the tree, and any that match another instance in the tree are replaced with a symbolic link to the other copy.  It's not very complicated code to deal with, but it doesn't save much in disk space or install time in the experiments we ran, and it resulted in a package tree with symbolic links that were somewhat haphazard for a human who might go poking around in node_modules.

It may be worth exploring again in the context of a massive monorepo, if we want to get the benefits of aggressive deduplication _without_ dependency isolation.

For that reason, I'd suggest the opt-in here be a `--mode=(default|isolated|...)` rather than a boolean, so that we can expand the set of reification modes in the future if we choose.

Reviewer 2: In fact, we're likely to eventually provide a `--mode=importmap` reification strategy, which would potentially deliver the value of `--mode=isolated`, but without any symlinks at all.

Reviewer: I added some bikeshedding comments on this.

These duplications come with the following cost:

- Performance degradation of the installation phase
- Inefficient disk usage
- Performance and memory cost at runtime as Node.js sees duplicated modules as two different modules.

### Tools crawling the node_modules folders

Certain tools implement their own module resolution algorithm instead of the one provided by Node.js. One of the motivation for a tool to implement its own resolution algorithm is that it can add more features to it.

Crawling the `node_modules` folder is such a feature used by certain build tools. This feature means that the mere presence of a package in a `node_modules` folder will have an impact on the output of a workspace. This means that any modification of a project's `node_modules` folder is possibly a breaking change to every workspaces in this project, regardless of the dependency graph.

For example, the TypeScript compiler by default includes to its compilation _every_ package in the folder `node_modules/@types` without any explicit import statement needed.

## Rationale

> The [import-maps](https://github.com/WICG/import-maps) standard is a very good tool to communicate the dependency graph to Node.js. But because this standard is not yet implemented in Node.js, npm needs an alternative way to solve the problems stated earlier.

Reviewer 3: ```suggestion
The [import-maps](https://github.com/WICG/import-maps) draft standard is a very good tool to communicate the dependency graph to Node.js. But because this feature is not yet implemented in Node.js, npm needs an alternative way to solve the problems stated earlier.
```
I believe it would require more than one browser implementation to be an actual standard

The goal is to provide an accurate dependency graph to Node.js while still relying on Node.js current module resolution algorithm.

> We deem valuable to invest the time implementing the `isolated-mode` now knowing that import-maps are coming because most of the implementation will be re-usable to implement a `import-maps-mode`.

Reviewer 3: ```suggestion
We deem it valuable to invest the time implementing `isolated-mode` now, knowing that import-maps are coming, because most of the implementation will be re-usable to implement an `import-maps-mode`.
```

> There are already two competing solutions out there which implement a `isolated-mode` ([pnpm](https://pnpm.io) and [yarn](https://yarnpkg.com)). We decided to choose the pnpm approach for the following reasons:

Reviewer 3: ```suggestion
There are already two competing solutions out there which implement an `isolated-mode`: ([pnpm](https://pnpm.io) and [yarn](https://yarnpkg.com)). We decided to choose the pnpm approach for the following reasons:
```

- _Works with current ecosystem_, the pnpm `isolated-mode` does not require any modification to Node.js or to the various build tools.
- _Battle tested_, Microsoft has successfully used pnpm to manage large monorepo for years.

> - _Recommended by Node.js_, this approach is actually the recommended approach by [the Node.js documentation](https://nodejs.org/api/modules.html#modules_addenda_package_manager_tips).

Reviewer 3: This link doesn’t go anywhere relevant - i was very shocked to read the implication that node made any official recommendation here. It’d be great to track down this content, or remove this bullet point if it can’t be located.

Author: the doc changed, this is the new link, I will update it, in the meantime the link is this: https://nodejs.org/api/modules.html#modules_package_manager_tips

Reviewer 3: oof, ok thanks - it even suggests using NODE_PATH, which is SUPER deprecated and not supported at all in ESM.

Author: While the NODE_PATH advice may be outdated, the advice related symlinks is probably not, the reasons for writing this in the NodeJs doc (presumably) still stand.

Reviewer 3: I suspect the reasons are that the author preferred this approach, and nobody reviewing the PR noticed that it’d be inappropriate for node to make such statements ¯\\\_(ツ)\_/¯

Reviewer 2: I can confidently report that yes, the author preferred that approach. :joy:

https://github.com/nodejs/node/commit/ff39d413a3c1#diff-5ae798d70a8dbd897d490bd7b453d5731ad3b49e416def53ee2ef0d11ca320b3R274

Additionally, once a module is resolved, the resolution algorithm calls 'realpath()' on the result. This means that the resolution algorithm always returns a real path. This allows to setup an arbitrary complex dependency graph while making sure Node.js does not create more than one instance of a given module.

## Detailed Explanation

### Structure

In the workspaces:

- Each workspace has its own `node_modules` folder.
- Each `node_modules` folder contain symlinks.
- The symlinks in a `node_modules` folder correspond to the matching workspace's dependencies.
- Each symlink is given the same name as the dependency it represents.
- Each symlink points to a location in the package store where the corresponding package is installed.

In the package store:

- The package store is a folder stored in the root/project level `node_modules` folder.
- The package store is named `.npm`.
- The package store contains folders, one for each dependency installed. These dependencies can be direct workspace dependencies or transitive dependencies.

> - These folders have a name containing the name of the package, its version and a hash of its content plus the hash of its dependencies.

Reviewer 3: what exactly is “its dependencies”? If it’s already hashing the content why is an additional hash needed?

Author: A package can be installed more than once in the store, this happens when a package has a peer dependency and is depended upon by two packages which provide a different version of this peer dependency (see one example in the example section). In this scenario, the folder of the package in the store should be different for the two different installations of that package. These packages will have the same content but different node_module folder, using a hash containing the dependencies of a package will make sure that these packages will not have a name collision.

Reviewer 3: If it’s the same version, then it has the same dependencies (since that’s the “dependencies” field in package.json), although it might have different dependents. If you mean it’s a hash of the node_modules folder, which might differ, that makes sense.

Author: In "dependencies" I really meant "dependencies+peerDependencies" but we really only need peerDependencies, I will change that.

Author: updated now, does it make more sense now?

> - These folders have a name containing the name of the package, its version and a hash of its content plus the hash of its resolved peer-dependencies.

Reviewer 2: ```suggestion
- These folders have a name containing the name of the package, its version and a hash of its content plus the hash of its resolved peer dependency set.
```

Reviewer 2: With a dependency graph like:

```
x ->PEER(y)
y ->PEER(z)
z ->()
```

You effectively have to treat _both_ `y` and `z` as the "peer set" of `x`, because if `x` depends on something with a conflicting peerDep on a different version of `z`, then that package cannot be installed alongside `y`.

- These folders contain a `node_modules` folder.
- These `node_modules` folders contain multiple folders, one for each dependency of that package and one for the package itself.
- All these folders are named based on their respective package names.
- The folder representing the package itself contains the code of that package. 
- The folders representation the dependencies of the package are symlinks to the corresponding package in the store.
- The dependencies of each packages are installed exactly the same way as workspaces' dependencies are installed.

This structure may seem more complicated than necessary. The reasons for this complexity are:

- to make it possible for a package to import itself using its own name. For example, the following code `require.resolve("foo");` in `node_modules/.npm/foo@1.0.0-1234/node_modules/foo/bar.js` will correctly return `node_modules/.npm/foo@1.0.0-1234/node_modules/foo/index.js`.
- to prevent creating circular symlinks, even when packages have circular dependencies.

> #### Peer dependencies

Reviewer 2: I think we can probably avoid a lot of this peerDep/edge case analysis if we specify that we are going to implement the isolation mode as a post-buildIdealTree conversion step.

That is, even though there are dependency graph shapes that may be enabled by isolated-mode, which are not enabled by hoisted mode, some some of these may be desirable, we are going to instead favor interoperability and reducing the degree of change in the dependency contracts enforced by npm.

So, whereas hoisted mode does the following:

1. Calculate hoisted ideal tree to satisfy dependency graph.
2. Reify ideal tree.
3. Save ideal tree to package-lock.json

instead, npm will:

1. Calculate hoisted ideal tree to satisfy dependency graph.
2. Capture hoisted tree lockfile metadata.
3. Convert hoisted tree into isolated tree.
4. Reify isolated tree.
5. Save hoisted tree lockfile metadata to package-lock.json

Thus, any situation not currently resolvable by hoisted-mode will not be possible with isolated-mode.  However, this ensures that any project may be reified using _either_ hoisted or isolated mode, or switched between them, without any changes to the project lockfile.

Anything that causes an `ERESOLVE` error in hoisted mode will also cause an `ERESOLVE` error in isolated mode.

Then the task is to document how that conversion process is accomplished.  Imo, that ought to end up in this RFC document, even if it happens after some implementation work has been done (which is probably a good idea, since we're bound to find some edge cases we haven't thought of!)

                     │                                │
                     │                                └─> B ( synlink to ../../../node_modules/.npm/B@1.0.0-0d98ab/node_modules/B )
                     │
                     └─> bar / ─┬─> [ content of workspace bar ]
                                │
                                └───> node_modules / ─┬─> A ( symlink to ../../../node_modules/.npm/A@1.0.0+B@2.0.0-66fe689/node_modules/A )
                                                      │
                                                      └─> B ( synlink to ../../../node_modules/.npm/B@2.0.0-a2ia56/node_modules/B )

```

## Performance

Compared to the current npm installation strategy, this proposal reduces package duplication, making the installation process faster. On a prototype, this installation strategy brought down the install time from 6 to 2 minutes on a large monorepo of 500+ workspaces.

## Configuration

The `mode` can be enabled by a CLI option when running `npm install` or by a setting in `.npmrc`.

## Compatibility

### Different versions of npm

> The `isolated-mode` will use the same lockfile as `hoisted` mode. Developers can use either mode without changing the lockfile. This makes `isolated-mode` backward compatible and easy to implement. The implementation of `isolated-mode` will reside in the reification, it will take the ideal-tree and apply it on disk using symlink instead of hoisting.

Reviewer 3: What happens if i use one reification, and then run npm install with another? Will it be able to trivially go back and forth without incident?

Author: it should be able to do the right thing, meaning change from hoisted to isolated and back with no leftover from the previous installation.

### Projects not using workspaces

> Non-workspace projects _will_ be able to use the `isolated-mode` reification. This is useful as a library author, using `isolated-mode` gives me the guarantee that my library will work as intended when consumed by projects which use `isolated-mode`.

Reviewer 2: ```suggestion
Non-workspace projects _will_ be able to use the `isolated-mode` reification. This is useful as a library author, using `isolated-mode` gives me the guarantee that my library will work as intended, without unexpected run-time errors caused by undeclared dependencies.
```

  - Junctions have to be representated by an absolute path, this means that junctions cannot be committed to git or packed into a package.
  - Symlinks can only be created in elevated shell [or when Windows is in "developer mode"](https://blogs.windows.com/windowsdeveloper/2016/12/02/symlinks-windows-10/#LCiVBWTgQF5s7fmL.97).

  - answer: junctions by default. support for symlinks can be added later as an opt-in if we see value for it.

- How much community code will break when the system forbids access to undeclared dependencies? In other words, how much code needs to be fixed to work properly in `isolated-mode`?
  - Regarding packages with missing dependencies in the `package.json` file. In many cases, package owners are willing to fix such issues. If there are a significant number of issues encountered around this, we can likely expect the of packages with missing dependencies is expected to drop fast after npm release `isolated-mode`. We may want to later add a feature to npm which allows users to locally declare dependencies on behalf of packages as a stop-gap, if existing solutions to this are not enough.

>   - There don't seem to be any packages out there that depend on the way hoisting works as a feature. If this is the case, it would be easy to argue that these packages should only rely on the npm and the Node.js contracts/APIs and not on implementation details like hoisting.

Reviewer 3: This remains inaccurate; there are such packages.

  - Some dev-environments don't support symlinks.
    - AWS Lambdas -> a repository can a be installed with isolated mode locally and on CI but then deployed in hoisted mode.
    - React native -> There are plugins existing to make react-native work with symlinks

>   - If a package is missing a dependency, it can be temporarily fixed (while waiting for the package owner to fix this issue) by declaring this missing dependency as top level dependency of the repository.

Reviewer 3: ```suggestion
  - If a package is missing a dependency, it can be temporarily fixed (while waiting for the package owner to fix this issue, or if it’s working as designed) by declaring this missing dependency as top level dependency of the repository.
```
